caffe2/operators/order_switch_ops_cudnn.cc

```diff
@@ -0,0 +1,160 @@
+#include "caffe2/operators/order_switch_ops.h"
+
+#include <algorithm>
+#include <functional>
+#include <vector>
+
+#include "caffe2/core/context_gpu.h"
+#include "caffe2/core/cudnn_wrappers.h"
+#include "caffe2/core/types.h"
+
+namespace caffe2 {
+
+namespace {
+
+class CuDNNOrderSwithOpBase : public Operator<CUDAContext> {
+ public:
+  USE_OPERATOR_FUNCTIONS(CUDAContext);
+
+  CuDNNOrderSwithOpBase(const OperatorDef& operator_def, Workspace* ws)
+      : Operator<CUDAContext>(operator_def, ws), cudnn_wrapper_(&context_) {
+    CUDNN_ENFORCE(cudnnCreateTensorDescriptor(&X_desc_));
+    CUDNN_ENFORCE(cudnnCreateTensorDescriptor(&Y_desc_));
+  }
+
+  virtual ~CuDNNOrderSwithOpBase() {
+    CUDNN_ENFORCE(cudnnDestroyTensorDescriptor(X_desc_));
+    CUDNN_ENFORCE(cudnnDestroyTensorDescriptor(Y_desc_));
+  }
+
+ protected:
+  void SetTensorDescriptor(
+      const cudnnDataType_t data_type,
+      const StorageOrder order,
+      const std::vector<int>& data_dims,
+      cudnnTensorDescriptor_t data_desc) const {
+    const int ndim = data_dims.size();
+    const int N = data_dims[0];
+    const int C = order == StorageOrder::NCHW ? data_dims[1] : data_dims.back();
+    if (ndim == 3) {
+      const int H = 1;
+      const int W = order == StorageOrder::NCHW ? data_dims[2] : data_dims[1];
+      CUDNN_ENFORCE(cudnnSetTensor4dDescriptor(
+          data_desc, GetCudnnTensorFormat(order), data_type, N, C, H, W));
+    } else if (ndim == 4) {
+      const int H = order == StorageOrder::NCHW ? data_dims[2] : data_dims[1];
+      const int W = order == StorageOrder::NCHW ? data_dims[3] : data_dims[2];
+      CUDNN_ENFORCE(cudnnSetTensor4dDescriptor(
+          data_desc, GetCudnnTensorFormat(order), data_type, N, C, H, W));
+    } else {
```

Comment:
shall we turn this to else if (ndim == 5) and add else to throw the error if it's not supported by CuDNN?

Reply:
It is fine to combine the dims greater than 5 together. It may not be as fast as 5d, but there will be almost no such cases. So I think it's fine to keep it as that.